presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushdownSubfields.java

```diff
@@ -556,42 +582,186 @@ private static NestedField nestedField(String name)
-                this.connectorSession = requireNonNull(connectorSession, "connectorSession is null");
+                this.connectorSession = connectorSession;
+                this.isPushDownSubfieldsFromLambdasEnabled = isPushDownSubfieldsFromLambdasEnabled;
+                this.functionAndTypeManager = functionAndTypeManager;
             }
 
             @Override
             public Void visitCall(CallExpression call, Context context)
             {
-                if (!functionResolution.isSubscriptFunction(call.getFunctionHandle())) {
-                    call.getArguments().forEach(argument -> argument.accept(this, context));
+                ComplexTypeFunctionDescriptor functionDescriptor = functionAndTypeManager.getFunctionMetadata(call.getFunctionHandle()).getDescriptor();
+                if (functionResolution.isSubscriptFunction(call.getFunctionHandle()) ||
+                        functionAndTypeManager.getFunctionMetadata(call.getFunctionHandle()).getName().equals(ELEMENT_AT)) {
+                    Optional<Subfield> subfield = toSubfield(call, functionResolution, expressionOptimizer, connectorSession, functionAndTypeManager);
+                    if (isPruningLambdaSubfieldsPossible(context)) {
+                        addRequiredLambdaSubfields(context, subfield);
+                        return null;
+                    }
+                    if (subfield.isPresent()) {
+                        context.subfields.add(subfield.get());
+                    }
+                    else {
+                        call.getArguments().forEach(argument -> argument.accept(this, context));
+                    }
                     return null;
                 }
+                if (isPushDownSubfieldsFromLambdasEnabled) {
+                    Optional<Set<Subfield>> lambdaSubfieldsOriginal = context.getLambdaSubfields();
+                    if ((functionDescriptor.isAccessingInputValues() && functionDescriptor.getLambdaDescriptors().isEmpty())
+                            || (functionDescriptor.isTransformingInput()) && !functionDescriptor.getOutputToInputTransformationFunction().isPresent()) {
+                        // If function internally accesses the input values we cannot prune any unaccessed subfields since we do not know what subfields function accessed.
+                        // If function transforms the shape of the input map or array argument and the form of such transformation was not supplied in
+                        // getOutputToInputTransformationFunction() then we have to discard all collected accessed subfields in lambdas of outer calls because we do not know how
+                        // to attribute those subfields to the input argument.
+                        // Additionally, we have to stop collecting lambda subfields in both cases until we find another function that does not return input values.
```

Comment:
what does "until we find another function that does not return input values" mean here.

Reply:
We push down lambda subfields only if we collected them. 
Function that does not return input values triggers the collecting of the lambda subfield. 
Consider the example 
```
CREATE TABLE orders (order_data ROW(ds VARCHAR, price BIGINT, quantity BIGINT, category VARCHAR))

SELECT
    ANY_MATCH(
        FILTER(
            TRANSFORM(order_data, CAST(ROW(ds, category) AS ROW(date, cat))),
            date >= '2023-01-01' AND date <= '2023-01-31'
        ),
        cat -> cat = 'FOOD'
    )
```
Here, `ANY_MATCH` does not return the value, and thus triggers the collecting of the lambda subfield. While traversing the inner calls, we collected the  `[*].date` and `[*].cat` subfields. However, as of now we do not have transformation between the input and the output of the `TRANSFORM` function and thus we cannot attribute those subfields to the input argument. Therefore, we have to drop all of them and stop collecting the lambda subfields (in particular, because we know that subfields were accessed) until we found another  function that does not return input values which will trigger the collecting of the lambda subfields again. 
If we did not find such function, then we will not pushdown any subfields which means file format reader will include all of them.

Reply:
Rephrased the comment `until we find another function that does not return input values` does not appeare there.